app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt

```diff
@@ -0,0 +1,224 @@
+        )
+
+        compositeDisposable.add(
+                highlightPublisher
+                        .debounce(INPUT_DEBOUNCE_MS, TimeUnit.MILLISECONDS)
+                        .map {
+                            parser.parse(lang, it.toString())
+                        }
+                        .subscribeOn(Schedulers.computation())
+                        .observeOn(Schedulers.computation())
+                        .subscribe(spanPublisher::onNext)
+        )
+
+        addTextChangedListener(this)
+        afterTextChanged(editableText)
+    }
+
+    override fun onDetachedFromWindow() {
+        removeTextChangedListener(this)
+        compositeDisposable.dispose()
+        scrollDisposable?.dispose()
+        super.onDetachedFromWindow()
+    }
+
+    var theme : CodeTheme = DefaultTheme
+        set(value) {
+            field = value
+            setBackgroundColor(value.background)
+            setTextColor(value.syntax.plain)
+            lineNumbersBackgroundPaint.color = value.lineNumberBackground
+            lineNumbersStrokePaint.color = value.lineNumberStroke
+            lineNumbersTextPaint.color = value.lineNumberText
+            selectedLinePaint.color = value.selectedLineBackground
+
+            afterTextChanged(editableText) // refresh highlight
+        }
+
+    private val bufferRect = Rect()
+
+    var lines: List<String> = emptyList()
+    var linesWithNumbers: List<Int> = emptyList()
+
+    override fun onDraw(canvas: Canvas) {
+        val lineNumbersOffset = lineNumbersTextPaint.measureText(lineCount.toString()).toInt() + 2 * LINE_NUMBERS_MARGIN
+
+        setPadding(lineNumbersOffset + LINE_NUMBERS_MARGIN, paddingTop, paddingRight, paddingBottom)
+        canvas.drawRect(0f, 0f, lineNumbersOffset.toFloat(), height.toFloat(), lineNumbersBackgroundPaint) // line numbers bg
+        canvas.drawLine(lineNumbersOffset.toFloat(), 0f, lineNumbersOffset.toFloat(), height.toFloat(), lineNumbersStrokePaint) // line numbers stroke
+
+        layout?.let { layout ->
+            var pos = 0
+            linesWithNumbers = lines.map {
+                val line = layout.getLineForOffset(pos)
+                pos += it.length + 1
+                line
+            }
+
+            linesWithNumbers.forEachIndexed { lineNumber, line ->
+                val y = getLineBounds(line, bufferRect)
+                canvas.drawText(lineNumber.toString(), lineNumbersOffset.toFloat() - LINE_NUMBERS_MARGIN, y.toFloat(), lineNumbersTextPaint)
+            }
+
+            val cursorPosition = selectionStart
+            if (cursorPosition > 0) {
+                val selectedLine = layout.getLineForOffset(cursorPosition)
+                var lineStart = selectedLine
+                var lineEnd = selectedLine + 1
+
+                while (!linesWithNumbers.contains(lineStart) && lineStart > 0) lineStart--
+                while (!linesWithNumbers.contains(lineEnd) && lineEnd < lineCount) lineEnd++
```

Comment:
what do this code? Please, extract it to the function. Can we optimize it (use hashset for example)?

Reply:
This code looks for upper and lover bound of text line. Usually text lines don't take more than 5 editor lines so there is no need in additional structures.